hack/toolkit-container-image.sh

```diff
@@ -13,11 +13,50 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-# Currently, this is hard-coded to the sha corresponding to the 1.18.0-rc.1 release
-# of the nvidia-container-toolkit. In the future, we should determine the exact
-# commit SHA corresponding to the version of the go module for the dependency
-# on `github.com/NVIDIA/nvidia-container-toolkit` -— whether it is a
-# tagged release or a pseudo-version -- and return that SHA instead.
-TOOLKIT_VERSION_SHA="4f98c014bfa1222a0b1dda34ec815f5ecf87c971"
-
-echo ghcr.io/nvidia/container-toolkit:${TOOLKIT_VERSION_SHA:0:8}
+# This script dynamically determines the nvidia-container-toolkit image to use
+# based on the version specified in go.mod. It:
+# 1. Uses go list to get the resolved version from go.mod
+# 2. Extracts SHA from pseudo-versions directly
+# 3. For tagged versions, resolves to commit SHA using git ls-remote
+#
+# The output is used in the Dockerfile to pull nvidia-cdi-hook binary from
+# the toolkit container's /artifacts/rpm/usr/bin/ directory.
+
+set -euo pipefail
+
+# Get the resolved version from go.mod using go tooling
+TOOLKIT_VERSION=$(go list -m -f '{{.Version}}' github.com/NVIDIA/nvidia-container-toolkit 2>/dev/null || true)
+
+if [ -z "${TOOLKIT_VERSION}" ]; then
+    echo "Error: Could not find nvidia-container-toolkit dependency" >&2
+    exit 1
+fi
+
+# Determine the commit SHA based on the version format
+TOOLKIT_VERSION_SHA=""
+
+# Check if this is a pseudo-version (format: v0.0.0-TIMESTAMP-SHA)
+if [[ "${TOOLKIT_VERSION}" =~ ^v[0-9]+\.[0-9]+\.[0-9]+-[0-9]{14}-([a-f0-9]{12})$ ]]; then
+    # Extract SHA from pseudo-version
+    TOOLKIT_VERSION_SHA="${BASH_REMATCH[1]}"
+elif [[ "${TOOLKIT_VERSION}" =~ ^v[0-9]+\.[0-9]+\.[0-9]+(-[a-zA-Z0-9.]+)?$ ]]; then
+    # This is a tagged release, fetch the SHA from GitHub
+    # Use git ls-remote with ^{} to dereference the tag and get the commit SHA
+    TOOLKIT_VERSION_SHA=$(git ls-remote --tags https://github.com/NVIDIA/nvidia-container-toolkit.git "refs/tags/${TOOLKIT_VERSION}^{}" 2>/dev/null | awk '{print $1}' || true)
+    
+    if [ -z "${TOOLKIT_VERSION_SHA}" ]; then
+        echo "Error: Could not resolve tag ${TOOLKIT_VERSION} to a commit SHA" >&2
+        exit 1
+    fi
+else
+    echo "Error: Unexpected version format: ${TOOLKIT_VERSION}" >&2
+    exit 1
+fi
```

Comment:
Does it make sense to construct / use the ghcr.io image only when a pseudo-version is present, and otherwise use the released version from nvcr.io?